life_expectancy/main.py

```diff
@@ -3,31 +3,36 @@
 
 import pandas as pd
 
-from life_expectancy.cleaning import clean_data
-from life_expectancy.loading_saving import load_data, save_data
+from life_expectancy.cleaning import DataCleaner
+from life_expectancy.loading_saving import DataLoader, save_data
 
 PARENT_PATH = Path(__file__).parent
 FILE_PATH = PARENT_PATH / 'data'
-INPUT_FILE_NAME = 'eu_life_expectancy_raw.tsv'
 OUTPUT_FILE_NAME = 'pt_life_expectancy.csv'
-INPUT_FILE_PATH = FILE_PATH / INPUT_FILE_NAME
 OUTPUT_FILE_PATH = FILE_PATH / OUTPUT_FILE_NAME
 
-def main(region: str) -> pd.DataFrame:
+
+def main(file_format: str, region: str) -> pd.DataFrame:
     """
     Main function.
     Method that loads, cleans and saves data.
 
     Args:
+    file_format [str]: file format (e.g. "csv", "json")
     region [str]: region to filter data.
     """
-    data_df = load_data(INPUT_FILE_PATH)
-    clean_df = clean_data(data_df, region)
-    save_data(clean_df, OUTPUT_FILE_PATH)
-    return clean_df
+
+    data_df = DataLoader(file_format).load_data()
+
+    cleaned_df = DataCleaner(file_format).clean_data(data_df, region)
+
```

Comment:
**Suggestion**: One problem I can see here is that you have to define the logic to select the correct treatment in 2 places (inside DataLoader and DataCleaner). If you add another compenent to this pipeline, you will need to add logic inside a 3rd object, and so forth. You are really not taking advantage of the Strategy Pattern as the number of places you need to change increase with the number of components.

My suggestion is for you to rearrange your code so that the logic to define the file type is done inside `main`. Yes, I know this would invalidade some of my previous suggestions... 😬

But it would look something like this:

```suggestion
    if file_path.suffix == ".tsv":
        loader, cleaner = DataLoaderTSV(), DataCleanerTSV()
    elif file_path.suffix == ".json":
        loader, cleaner = DataLoaderJSON(), DataCleanerJSON()

    df = loader.load_data()
    cleaned_dv = cleaner.clean_data(df, region)
```

It's like lego. You submodule just hold the your lego pieces (classes and functions) and your main.py is where the logic is applied and the pieces are put together. This is works for small repos such as these and it's very easy to maintain.

Reply:
I changed the code according your suggestions! Thank you!